### PR TITLE
Closes #723 added LongInterval

### LINKED ISSUE #723: Implement LongInterval
Today we have IntInterval.  It would be useful to have LongInterval as well.  I don't know that we need more than just IntInterval and LongInterval, so I would hold off trying to code generate all of the primitive type versions of interval.

### PR DESCRIPTION
**Pull request by vmzakharov**
`LongInterval` supports elements of type 'long' but is otherwise consistent with other primitive lists and is limited to an int size.

Signed-off-by: vmzakharov <zakharov.vladimir.m@gmail.com>

### CODE REVIEW

**Comment by nikhilnanivadekar** [comment]
@mohrezaei can you do a review please? I did a rudimentary review. Would like your final review. Thanks!

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java [hunk 1]
```diff
@@ -0,0 +1,1092 @@
...
+        {
+            sum += this.get(i) * list.get(i);
+        }
+        return sum;
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return this.size() == 0;
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:565]: isn't this always false?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java [hunk 2]
```diff
@@ -0,0 +1,1092 @@
+/*
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:1]: [N.B. comment applies to whole class, so placed up here]
Why is the looping in this class so inconsistent? Sometimes it's fully expanded (e.g. `forEach`, `equals`, etc). Sometimes, it's using a for-loop with `this.get(i)`. Sometimes, it's using `forEach*` with a lambda. And yet other places, a `LongIterator` is used. It's like 4 different people wrote this class 😄

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java [hunk 3]
```diff
@@ -0,0 +1,1092 @@
...
+    @Override
+    public LazyLongIterable asReversed()
+    {
+        return ReverseLongIterable.adapt(this);
+    }
+
+    @Override
+    public long sum()
+    {
+        long sum = 0L;
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:827]: does the formula based implementation not work for longs? overflow?

Comment (vmzakharov) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:827]: Should work, will fix

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java (hunk 1 above)
Comment (vmzakharov) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:565]: Same issue in `IntInterval` (which I effectively cloned), can change implementation/add tests to both or have a separate PR to clean up this and other implementation issues. What do you suggest?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java (hunk 2 above)
Comment (vmzakharov) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:1]: Same as above - I followed the implementation pattern of `IntInterval` - one reason didn't want to diverge because it would make it harder to turn it into a code generated template if we decide to do so in the future.   My preference is to leave it as it and have a separate PR to clean up both classes (and keep them consistent). Please let me know what you think.

Comment (mohrezaei) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:1]: My preference would be to have a clean implementation here and another PR to clean up IntInterval. Others are welcome to chime in. @donraab @nikhilnanivadekar @motlin ?

Comment (nikhilnanivadekar) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:1]: I am more of let’s get this one in first and then clean up one or both. I am waiting for these PRs to be closer to release 10.3.0. So the fastest way to that is more preferable for me.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java (hunk 1 above)
Comment (nikhilnanivadekar) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:565]: For my understanding, is the proposal to return `false` instead of doing the check? Like Moh says it will always be `false` but this implementation is not incorrect or did I miss something? Interval always makes me think twice about these terminating methods because Interval is a mix of eager and lazy functionality.

Comment (mohrezaei) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:565]: yes, the proposal is to return false. An implementation that computes PI to 700 digits and then returns `lastDigit == 0` would also work (the 700th digit of PI is 5) 😛 .

Comment (nikhilnanivadekar) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:565]: LOL 😂

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java (hunk 2 above)
Comment (vmzakharov) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/primitive/LongInterval.java:1]: I have fixed `sum()` and added more tests for it (specifically to address the concern about overflow during calculation), and have added a test for `isEmpty()`. My preference now would be to get `LongInterval` in as is and open a new issue for `LongInterval` and `IntInterval` code clean up (happy to work on it as the potential is certainly there!)
